File: adventJs/2024/10-the-elfish-assembler/__main__.py
def compile(instructions):
    registry = {}
    index = 0

    def createReg(reg):
        registry.setdefault(reg, 0)

    def mov(value, reg):
        createReg(reg)
        try:
            registry[reg] = int(value)
        except ValueError:
            registry[reg] = registry[value]
        nonlocal index
        index += 1

    def inc(reg):
        createReg(reg)
        registry[reg] += 1
        nonlocal index
        index += 1

    def dec(reg):
        createReg(reg)
        registry[reg] -= 1
        nonlocal index
        index += 1

    def jmp(reg, jump):
        createReg(reg)
        nonlocal index
        index = int(jump) if registry[reg] == 0 else index + 1

    actions = {"MOV": mov, "INC": inc, "DEC": dec, "JMP": jmp}

    while index < len(instructions):
        action, *instruction = instructions[index].split(" ")
        print({"index": index, "reg": registry, "action": action})
        actions[action](*instruction)

    return registry["A"]

File: adventJs/2024/10-the-elfish-assembler/main.js
/**
 * @param {string[]} instructions - The instructions to execute
 * @returns {number} The value of the register A
 */
function compile(instructions) {
  let registry = {}
  let index = 0

  const createReg = (reg) => { if (registry[reg] === undefined) registry[reg] = 0 }

  const actions = {
    'MOV': ([value, reg]) => {
      createReg(reg)
      isNaN(value) ? registry[reg] = registry[value] : registry[reg] = +value
      index++
    },
    'INC': ([reg]) => {
      createReg(reg)
      registry[reg] += 1
      index++
    },
    'DEC': ([reg]) => {
      createReg(reg)
      registry[reg] -= 1
      index++
    },
    'JMP': ([reg, jump]) => {
      createReg(reg)
      registry[reg] === 0 ? index = +jump : index++
    }
  }

  while (index < instructions.length) {
    const [action, ...instruction] = instructions[index].split(' ')
    actions[action](instruction)
  }

  return registry['A']
}
